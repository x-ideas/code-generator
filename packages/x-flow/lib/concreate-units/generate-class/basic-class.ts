export class CBasicModel {
  #parsedErrorTips: string[] = [];

  /**
   * 解析过程是否成功
   * @returns {boolean}
   * @memberof CBasicModel
   */
  isParsedCorrect(): boolean {
    // 暂时未提供，方案待确定
    return true;
    // return this.#parsedErrorTips.length === 0;
  }

  /**
   * 新增解析过程的错误提示
   * @param {string} tip
   * @memberof CBasicModel
   */
  addParsedErrorTip(tip: string): void {
    this.#parsedErrorTips.push(tip);
  }
}
