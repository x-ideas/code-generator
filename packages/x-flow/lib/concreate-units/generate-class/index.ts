/* eslint-disable @typescript-eslint/ban-ts-comment */
/*
 * @name:
 * description:
 *  将interface生成前端class定义的FlowUnit
 */

import * as babelParser from '@babel/parser';
import lodash from 'lodash';

import { XFlowUnit } from '../../flow-unit';

import jscodeshift from 'jscodeshift';
import { Collection } from 'jscodeshift/src/Collection';
import * as K from 'jscodeshift/node_modules/ast-types/gen/kinds';
import { parserConfig } from '../../utils/jscodeshift-parser';

export class GenerateClassFlowUnit extends XFlowUnit {
  #convertedFromNiceFormat: boolean;
  constructor({ convertedFromNiceFormat = true }: { convertedFromNiceFormat: boolean }) {
    super();

    this.#convertedFromNiceFormat = convertedFromNiceFormat;
  }

  /**
   *
   * @private
   * @param {Collection<jscodeshift.>} collection
   * @returns {Collection<ASTNode>}
   * @memberof GenerateClassFlowUnit
   */
  private _changeInterfaceToClass(collection: Collection<jscodeshift.ASTNode>): Collection<jscodeshift.ASTNode> {
    const newCollection = jscodeshift('');
    const newProgram = newCollection.find(jscodeshift.Program);

    collection.find(jscodeshift.TSInterfaceDeclaration).forEach(path => {
      const classDeclaration = generateClassDeclaration(path, {
        convertedFromNiceFormat: this.#convertedFromNiceFormat,
      });
      newProgram.get('body').value.push(classDeclaration);
    });

    return newCollection;
  }

  async doWork(interfaceSources: string): Promise<string> {
    const collection: Collection<jscodeshift.ASTNode> = jscodeshift(interfaceSources, {
      parser: parserConfig(),
    });
    // const a = collection.find(jscodeshift.TSInterfaceDeclaration).get('body').get('body');
    // // console.log(collection.find(jscodeshift.TSInterfaceDeclaration).length);

    // // const ds = collection.find(jscodeshift.TSInterfaceDeclaration);
    // // const id = ds.get('id');
    // console.log(collection.find(jscodeshift.TSInterfaceDeclaration).get('id').get('name').value);

    return this._changeInterfaceToClass(collection).toSource();
  }
}

/**
 * 生成exporse装饰器
 * @param {jscodeshift.TSPropertySignature} property
 * @returns {jscodeshift.Decorator}
 */
function generateExposeDecorator(
  property: jscodeshift.TSPropertySignature,
  options: {
    convertedFromNiceFormat: boolean;
  }
): jscodeshift.Decorator {
  // 找到参数
  const argums: jscodeshift.ObjectExpression[] = [];
  if (jscodeshift.Identifier.check(property.key)) {
    // property

    const bName: string = options.convertedFromNiceFormat ? lodash.snakeCase(property.key.name) : property.key.name;

    const nameProperty = jscodeshift.property('init', jscodeshift.identifier('name'), jscodeshift.literal(bName));
    argums.push(jscodeshift.objectExpression([nameProperty]));
  }

  return jscodeshift.decorator(jscodeshift.callExpression(jscodeshift.identifier('Expose'), argums));
}

/**
 * 生成validate装饰器
 * TODO: 判断是否是可选的属性
 * @param {jscodeshift.TSPropertySignature} property
 * @returns {jscodeshift.Decorator}
 */
function generateValidateDecorator(property: jscodeshift.TSPropertySignature): jscodeshift.Decorator | null {
  // let name: string = '';
  // if (jscodeshift.Identifier.check(property.key)) {
  //   name = property.key.name;
  // }

  // if (!name) {
  //   return null;
  // }

  if (property.optional) {
    return null;
  }

  if (property.typeAnnotation && jscodeshift.TSTypeAnnotation.check(property.typeAnnotation)) {
    // TODO: 根据类型和name生成对应的call expression

    switch (property.typeAnnotation.typeAnnotation.type) {
      // number, enum当作number
      case 'TSNumberKeyword': {
        return jscodeshift.decorator(
          jscodeshift.callExpression(jscodeshift.identifier('ValidateValue'), [
            jscodeshift.callExpression(
              jscodeshift.identifier.from({
                comments: [jscodeshift.commentBlock('NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型')],
                name: 'getDefaultValidNum',
              }),
              []
            ),
          ])
        );

        break;
      }

      case 'TSStringKeyword': {
        // 判断是否为id // string;
        return jscodeshift.decorator(
          jscodeshift.callExpression(jscodeshift.identifier('ValidateValue'), [
            jscodeshift.callExpression(jscodeshift.identifier('getDefaultValidStr'), []),
          ])
        );
      }

      case 'TSBooleanKeyword':
        // boolean
        return jscodeshift.decorator(
          jscodeshift.callExpression(jscodeshift.identifier('ValidateValue'), [
            jscodeshift.callExpression(jscodeshift.identifier('getDefaultValidBoolean'), []),
          ])
        );

      case 'TSTypeReference':
        // 另一个类型定义(interface, class, enum),
        // NOTE: 通过前缀区分
        return jscodeshift.decorator(
          jscodeshift.callExpression(jscodeshift.identifier('Type'), [
            // @ts-ignore
            jscodeshift.arrowFunctionExpression([], jscodeshift.identifier(property.typeAnnotation.typeAnnotation.typeName?.name)),
          ])
        );

      case 'TSArrayType':
        {
          if (property.typeAnnotation.typeAnnotation.elementType.type === 'TSTypeReference') {
            const typeName = property.typeAnnotation.typeAnnotation.elementType.typeName;
            if (jscodeshift.Identifier.check(typeName)) {
              return jscodeshift.decorator(
                jscodeshift.callExpression(jscodeshift.identifier('Type'), [
                  jscodeshift.arrowFunctionExpression([], jscodeshift.identifier(typeName.name)),
                ])
              );
            }
          }
        }

        break;

      default:
        break;
    }
  }

  return null;
}

/**
 * 生成class property的初始化值
 * @param {jscodeshift.TSPropertySignature} property
 */
function generateClassPropertyInitValue(property: jscodeshift.TSPropertySignature): K.ExpressionKind | null {
  // 找到type类型
  const optional = property.optional;
  if (optional) {
    return null;
  }

  switch (property.typeAnnotation?.typeAnnotation.type) {
    case 'TSBooleanKeyword':
      return jscodeshift.literal(false);

    case 'TSNumberKeyword':
      return jscodeshift.literal(0);

    case 'TSStringKeyword': {
      // 是否是id
      jscodeshift.Identifier.assert(property.key);
      if (jscodeshift.Identifier.check(property.key)) {
        if (property.key.name.includes('id')) {
          return jscodeshift.callExpression(jscodeshift.identifier('getFrontId'), []);
        }
      }

      return jscodeshift.stringLiteral('');
    }

    case 'TSArrayType':
      return jscodeshift.arrayExpression([]);

    case 'TSTypeReference': {
      // TODO: 根据名字去判断
      return null;
    }

    default:
      return null;
  }
}

/**
 * 生成class property的标注
 * 如果property是可选的，那么会得到一个联合类型
 * @param {jscodeshift.TSParameterProperty} property
 * @returns {TSTypeAnnotationKind}
 */
function generateClassPropertyAnnotation(property: jscodeshift.TSPropertySignature): jscodeshift.TSTypeAnnotation | undefined {
  function _generate(typeAnnotation: K.TSTypeKind | K.TSTypeAnnotationKind, optional = false): jscodeshift.TSTypeAnnotation | undefined {
    switch (typeAnnotation.type) {
      case 'TSBooleanKeyword': {
        return optional
          ? jscodeshift.tsTypeAnnotation(jscodeshift.tsUnionType([jscodeshift.tsBooleanKeyword(), jscodeshift.tsUndefinedKeyword()]))
          : jscodeshift.tsTypeAnnotation(jscodeshift.tsBooleanKeyword());
      }
      case 'TSNumberKeyword': {
        return optional
          ? jscodeshift.tsTypeAnnotation(jscodeshift.tsUnionType([jscodeshift.tsNumberKeyword(), jscodeshift.tsUndefinedKeyword()]))
          : jscodeshift.tsTypeAnnotation(jscodeshift.tsNumberKeyword());
      }
      case 'TSStringKeyword': {
        return optional
          ? jscodeshift.tsTypeAnnotation(jscodeshift.tsUnionType([jscodeshift.tsStringKeyword(), jscodeshift.tsUndefinedKeyword()]))
          : jscodeshift.tsTypeAnnotation(jscodeshift.tsStringKeyword());
      }

      case 'TSArrayType': {
        return optional
          ? jscodeshift.tsTypeAnnotation(
              jscodeshift.tsUnionType([jscodeshift.tsArrayType(typeAnnotation.elementType), jscodeshift.tsUndefinedKeyword()])
            )
          : jscodeshift.tsTypeAnnotation(jscodeshift.tsArrayType(typeAnnotation.elementType));
      }

      case 'TSTypeReference': {
        return optional
          ? jscodeshift.tsTypeAnnotation(
              jscodeshift.tsUnionType([jscodeshift.tsTypeReference(typeAnnotation.typeName), jscodeshift.tsUndefinedKeyword()])
            )
          : jscodeshift.tsTypeAnnotation(jscodeshift.tsTypeReference(typeAnnotation.typeName));
      }
    }
    return undefined;
  }

  if (property.typeAnnotation) {
    switch (property.typeAnnotation.type) {
      case 'TSTypeAnnotation': {
        return _generate(property.typeAnnotation.typeAnnotation, property.optional);
        break;
      }

      default:
        break;
    }
  }

  return undefined;
}

/**
 * 生成class的属性定义
 * @param {jscodeshift.TSPropertySignature} property
 * @returns {(jscodeshift.ClassProperty | undefined)}
 */
function generateClassPropertyFromInterfaceProperty(
  property: jscodeshift.TSPropertySignature,
  options: {
    convertedFromNiceFormat: boolean;
  }
): jscodeshift.ClassProperty | null {
  if (jscodeshift.Identifier.check(property.key)) {
    const ccName = lodash.camelCase(property.key.name);

    const value = generateClassPropertyInitValue(property);
    const typeAnnotation = generateClassPropertyAnnotation(property);
    const cProperty = jscodeshift.classProperty(jscodeshift.identifier(ccName), value, typeAnnotation);
    cProperty.comments = property.comments;

    // @ts-ignore
    // NOTE: 定义居然没有
    cProperty.decorators = [generateExposeDecorator(property, options), generateValidateDecorator(property)].filter(Boolean);

    return cProperty;
  }
  return null;
}

/**
 * 生成class声明
 * @param {string} className
 * @param {jscodeshift.TSPropertySignature[]} properties
 * @returns {jscodeshift.ExportNamedDeclaration}
 */
function generateClassDeclaration(
  interfaceDeclaration: jscodeshift.ASTPath<jscodeshift.TSInterfaceDeclaration>,
  options: {
    convertedFromNiceFormat: boolean;
  }
): jscodeshift.ExportNamedDeclaration {
  const interfaceName: string = jscodeshift(interfaceDeclaration).get('id', 'name').value;
  const properties = jscodeshift(interfaceDeclaration).get('body', 'body').value;

  // @ts-ignore
  const cProperties = properties.map(prop => generateClassPropertyFromInterfaceProperty(prop, options));

  // 将第一个I替换成C
  return jscodeshift.exportNamedDeclaration(
    jscodeshift.classDeclaration(jscodeshift.identifier(`C${interfaceName.slice(1)}`), jscodeshift.classBody(cProperties))
  );
}
