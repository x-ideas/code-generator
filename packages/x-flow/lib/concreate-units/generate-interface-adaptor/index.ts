import { XFlowUnit } from '../../flow-unit';

import * as babelParser from '@babel/parser';
import lodash from 'lodash';

import jscodeshift, { ASTPath } from 'jscodeshift';
import { Collection } from 'jscodeshift/src/Collection';

interface IGenerateInterfaceAdaptorOptions {
  fromType: string;
  from: string;

  toType: string;
  // to: string;

  isConvertedFromFront: boolean;
}

export class GenerateInterfaceAdaptor extends XFlowUnit {
  _getFuncName(name: string, options: IGenerateInterfaceAdaptorOptions) {
    return `to${options.isConvertedFromFront ? 'B' : 'F'}${name}Adaptor`;
  }

  _generateInterfaceAdaptor(
    objectName: string,
    options: IGenerateInterfaceAdaptorOptions,
    properties: jscodeshift.TSPropertySignature[]
  ): jscodeshift.Property[] {
    const result: jscodeshift.Property[] = [];

    for (const property of properties) {
      if (jscodeshift.Identifier.check(property.key)) {
        const name = property.key.name;

        let isRef = false;
        let typeName = '';
        let isArray = false;

        if (jscodeshift.TSTypeReference.check(property.typeAnnotation?.typeAnnotation)) {
          isRef = true;
          if (jscodeshift.Identifier.check(property.typeAnnotation?.typeAnnotation.typeName))
            typeName = property.typeAnnotation?.typeAnnotation.typeName.name ?? '';
        } else if (jscodeshift.TSArrayType.check(property.typeAnnotation?.typeAnnotation)) {
          isArray = true;

          if (jscodeshift.TSTypeReference.check(property.typeAnnotation?.typeAnnotation.elementType)) {
            isRef = true;
            if (jscodeshift.Identifier.check(property.typeAnnotation?.typeAnnotation.elementType.typeName))
              typeName = property.typeAnnotation?.typeAnnotation.elementType.typeName.name ?? '';
          }
        }

        const memberExpression = jscodeshift.memberExpression(jscodeshift.identifier(objectName), jscodeshift.identifier(name));
        result.push(
          jscodeshift.property(
            'init',
            jscodeshift.identifier(options.isConvertedFromFront ? lodash.snakeCase(name) : lodash.camelCase(name)),
            isArray
              ? isRef
                ? jscodeshift.callExpression(memberExpression, [
                    jscodeshift.arrowFunctionExpression(
                      [jscodeshift.identifier('item')],
                      jscodeshift.callExpression(jscodeshift.identifier(this._getFuncName(typeName, options)), [jscodeshift.identifier('item')])
                    ),
                  ])
                : memberExpression
              : isRef
              ? jscodeshift.callExpression(jscodeshift.identifier(this._getFuncName(typeName, options)), [memberExpression])
              : memberExpression
          )
        );
      }
    }

    return result;
  }

  async doWork(options: IGenerateInterfaceAdaptorOptions): Promise<string[]> {
    const collection: Collection<jscodeshift.ASTNode> = jscodeshift(options.from, {
      parser: {
        parse(source: string) {
          return babelParser.parse(source, {
            sourceType: 'module',
            // 支持typescript, jsx
            plugins: [
              'estree',
              'typescript',
              [
                'decorators',
                {
                  decoratorsBeforeExport: true,
                },
              ],
              'exportDefaultFrom',
              'classProperties',
              'classPrivateProperties',
            ],
          });
        },
      },
    });

    const funcStr: string[] = [];
    collection.find(jscodeshift.TSInterfaceDeclaration).forEach((path: ASTPath<jscodeshift.TSInterfaceDeclaration>) => {
      if (jscodeshift.Identifier.check(path.node.id)) {
        const interfaceName = path.node.id.name;

        const interfaceProperties: jscodeshift.TSPropertySignature[] = path.node.body.body as jscodeshift.TSPropertySignature[];

        const func = jscodeshift.functionDeclaration(
          jscodeshift.identifier(this._getFuncName(interfaceName, options)),
          [
            jscodeshift.identifier.from({
              name: options.isConvertedFromFront ? 'fInfo' : 'bInfo',
              typeAnnotation: jscodeshift.tsTypeAnnotation(jscodeshift.tsTypeReference(jscodeshift.identifier(options.fromType))),
            }),
          ],
          jscodeshift.blockStatement([])
        );
        func.returnType = jscodeshift.tsTypeAnnotation(jscodeshift.tsTypeReference(jscodeshift.identifier(options.toType)));

        const result: Collection<any> = jscodeshift(func);

        const functionBody = func.body.body;

        functionBody.push(
          jscodeshift.returnStatement(
            jscodeshift.objectExpression(
              this._generateInterfaceAdaptor(options.isConvertedFromFront ? 'fInfo' : 'bInfo', options, interfaceProperties)
            )
          )
        );

        funcStr.push(result.toSource());
      }
    });

    return funcStr;
  }
}
