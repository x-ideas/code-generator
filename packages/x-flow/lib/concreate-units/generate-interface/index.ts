/* eslint-disable @typescript-eslint/ban-ts-comment */
/*
 * @name:
 * description:
 *  后台interface生成的处理逻辑
 */

import { InputData, JSONSchemaInput, quicktype, TypeScriptTargetLanguage } from 'quicktype-core';
import { JSONSchema4 } from 'json-schema';
import jscodeshift from 'jscodeshift';

import lodash from 'lodash';
import { XFlowUnit } from '../../flow-unit';
import { parserConfig } from '../../utils/jscodeshift-parser';

type TOption = {
  nicePropertyName: boolean;
  fInterfacePrefix?: string;
  bInterfacePrefix?: string;
};

/**
 * 将json scheme生成interface的处理流程
 * @export
 * @class BackEndInterfaceGenerateFlowUnit
 * @extends {XFlowUnit}
 */
export class InterfaceGenerateFlowUnit extends XFlowUnit {
  #options: TOption;

  constructor(options: TOption) {
    super();

    this.#options = options;
  }

  /**
   *
   * @param params topName不带IF,由nicePropertyName去添加
   */
  async doWork(params: { jsonSchema: JSONSchema4; topName: string }): Promise<string> {
    const inputData = new InputData();

    const source = { name: params.topName, schema: JSON.stringify(params.jsonSchema) };
    inputData.addSource('schema', source, () => new JSONSchemaInput(undefined));

    const { lines } = await quicktype({
      lang: new TypeScriptTargetLanguage(),
      inputData,

      // leadingComments: ['测试备注'],
      rendererOptions: {
        'just-types': 'true',
        // @ts-ignore
        'nice-property-names': this.#options.nicePropertyName ? 'true' : '',
        'acronym-style': 'camel',
      },
    });

    const prefix = this.#options.nicePropertyName ? this.#options.fInterfacePrefix ?? 'IF' : this.#options.bInterfacePrefix ?? 'IB';

    const replace1 = jscodeshift(lines.join('\n'), {
      parser: parserConfig(),
    })
      .find(jscodeshift.TSInterfaceDeclaration)
      .forEach(path => {
        //

        if (jscodeshift.Identifier.check(path.node.id)) {
          path.node.id.name = `${prefix}${path.node.id.name}`;
        }
      })
      .toSource();

    const replace2 = jscodeshift(replace1, {
      parser: parserConfig(),
    })
      .find(jscodeshift.TSTypeReference)
      .forEach(path => {
        //

        if (jscodeshift.Identifier.check(path.node.typeName)) {
          path.node.typeName.name = `${prefix}${path.node.typeName.name}`;
        }
      })
      .toSource();

    if (this.#options.nicePropertyName) {
      return replace2;
    } else {
      return replace2.replace(/export/g, '');
    }
  }
}
