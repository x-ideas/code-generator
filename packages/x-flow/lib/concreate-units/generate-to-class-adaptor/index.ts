/*
 * @name:
 * description:
 *  生成class的适配函数
 */
import jscodeshift from 'jscodeshift';

import { XFlowUnit } from '../../flow-unit';

export class GenerateToClassAdaptorFlowUnit extends XFlowUnit {
  private generateFunc(className: string) {
    const funcDeclaration = jscodeshift.functionDeclaration(
      jscodeshift.identifier(`to${className}AdaptorFunc`),
      [
        jscodeshift.identifier.from({
          name: 'bInfo',
          comments: [],
          typeAnnotation: jscodeshift.typeAnnotation(jscodeshift.anyTypeAnnotation()),
        }),
      ],
      jscodeshift.blockStatement([
        jscodeshift.returnStatement(
          jscodeshift.callExpression(
            jscodeshift.memberExpression(
              jscodeshift.callExpression(jscodeshift.identifier('plainToClass'), [
                jscodeshift.identifier(className),
                jscodeshift.identifier('bInfo'),
                jscodeshift.objectExpression([
                  jscodeshift.property('init', jscodeshift.identifier('excludeExtraneousValues'), jscodeshift.literal(true)),
                  jscodeshift.property(
                    'init',
                    jscodeshift.identifier('groups'),
                    jscodeshift.arrayExpression([
                      jscodeshift.memberExpression(jscodeshift.identifier('EExportGroup'), jscodeshift.identifier('Always')),
                    ])
                  ),
                ]),
              ]),
              jscodeshift.identifier('filter')
            ),
            [
              jscodeshift.arrowFunctionExpression(
                [jscodeshift.identifier('item')],
                jscodeshift.callExpression(jscodeshift.memberExpression(jscodeshift.identifier('item'), jscodeshift.identifier('isValid')), [])
              ),
            ]
          )
        ),
      ])
    );

    // funcDeclaration.returnType = jscodeshift.typeAnnotation(jscodeshift.iterface(jscodeshift.identifier('dd')));
    return funcDeclaration;
  }

  async doWork(className: string) {
    const program = jscodeshift('').find(jscodeshift.Program);
    program.get('body').value.push(this.generateFunc(className));

    return {
      funcName: `to${className}AdaptorFunc`,
      func: program.toSource(),
    };
  }
}
