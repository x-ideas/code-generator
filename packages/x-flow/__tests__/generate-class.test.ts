import { GenerateClassFlowUnit } from '../lib/concreate-units/generate-class';

describe('测试 generate class', () => {
  test('一个interface', async () => {
    const str = `      /**
       * 进度
       */
      export interface IProgress {
          /**
           * 集货进度
           */
          collect_progress?: number;
          /**
           * 相对集货进度
           */
          collect_progress_relative?: number;
          /**
           * 校正集货数总和
           */
          collect_sum_correct?: number;
          /**
           * 待处理集货数总和
           */
          collect_sum_pending?: number;
          /**
           * 计划集货数总和
           */
          collect_sum_plan?: number;
          /**
           * 实际集货数总和
           */
          collect_sum_real?: number;
          /**
           * 出货进度
           */
          delivery_progress?: number;
          /**
           * 相对出货进度
           */
          delivery_progress_relative?: number;
          /**
           * 实际出货数总和
           */
          delivery_sum_real?: number;
          /**
           * 分货进度
           */
          distribute_progress?: number;
          /**
           * 相对分货进度
           */
          distribute_progress_relative?: number;
          /**
           * 实际分货数总和
           */
          distribute_sum_real?: number;
          /**
           * 波次创建进度
           */
          order_progress?: number;
          /**
           * 收货进度
           */
          receive_progress?: number;
          /**
           * 相对收货进度
           */
          receive_progress_relative?: number;
          /**
           * 实际入库数总和
           */
          receive_sum_real?: number;
      }`;

    const unit = new GenerateClassFlowUnit({ convertedFromNiceFormat: false });

    const result = await unit.doWork(str);

    expect(result).toMatchSnapshot();
  });
});
