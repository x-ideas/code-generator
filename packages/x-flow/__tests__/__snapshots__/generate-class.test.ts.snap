// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试 generate class 一个interface 1`] = `
"export class Crogress {
  /**
   * 集货进度
   */
  @Expose({
    name: \\"collect_progress\\"
  })
  @ValidateValue(/*NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型*/
  getDefaultValidNum())
  collectProgress: number | undefined;

  /**
   * 相对集货进度
   */
  @Expose({
    name: \\"collect_progress_relative\\"
  })
  @ValidateValue(/*NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型*/
  getDefaultValidNum())
  collectProgressRelative: number | undefined;

  /**
   * 校正集货数总和
   */
  @Expose({
    name: \\"collect_sum_correct\\"
  })
  @ValidateValue(/*NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型*/
  getDefaultValidNum())
  collectSumCorrect: number | undefined;

  /**
   * 待处理集货数总和
   */
  @Expose({
    name: \\"collect_sum_pending\\"
  })
  @ValidateValue(/*NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型*/
  getDefaultValidNum())
  collectSumPending: number | undefined;

  /**
   * 计划集货数总和
   */
  @Expose({
    name: \\"collect_sum_plan\\"
  })
  @ValidateValue(/*NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型*/
  getDefaultValidNum())
  collectSumPlan: number | undefined;

  /**
   * 实际集货数总和
   */
  @Expose({
    name: \\"collect_sum_real\\"
  })
  @ValidateValue(/*NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型*/
  getDefaultValidNum())
  collectSumReal: number | undefined;

  /**
   * 出货进度
   */
  @Expose({
    name: \\"delivery_progress\\"
  })
  @ValidateValue(/*NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型*/
  getDefaultValidNum())
  deliveryProgress: number | undefined;

  /**
   * 相对出货进度
   */
  @Expose({
    name: \\"delivery_progress_relative\\"
  })
  @ValidateValue(/*NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型*/
  getDefaultValidNum())
  deliveryProgressRelative: number | undefined;

  /**
   * 实际出货数总和
   */
  @Expose({
    name: \\"delivery_sum_real\\"
  })
  @ValidateValue(/*NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型*/
  getDefaultValidNum())
  deliverySumReal: number | undefined;

  /**
   * 分货进度
   */
  @Expose({
    name: \\"distribute_progress\\"
  })
  @ValidateValue(/*NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型*/
  getDefaultValidNum())
  distributeProgress: number | undefined;

  /**
   * 相对分货进度
   */
  @Expose({
    name: \\"distribute_progress_relative\\"
  })
  @ValidateValue(/*NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型*/
  getDefaultValidNum())
  distributeProgressRelative: number | undefined;

  /**
   * 实际分货数总和
   */
  @Expose({
    name: \\"distribute_sum_real\\"
  })
  @ValidateValue(/*NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型*/
  getDefaultValidNum())
  distributeSumReal: number | undefined;

  /**
   * 波次创建进度
   */
  @Expose({
    name: \\"order_progress\\"
  })
  @ValidateValue(/*NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型*/
  getDefaultValidNum())
  orderProgress: number | undefined;

  /**
   * 收货进度
   */
  @Expose({
    name: \\"receive_progress\\"
  })
  @ValidateValue(/*NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型*/
  getDefaultValidNum())
  receiveProgress: number | undefined;

  /**
   * 相对收货进度
   */
  @Expose({
    name: \\"receive_progress_relative\\"
  })
  @ValidateValue(/*NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型*/
  getDefaultValidNum())
  receiveProgressRelative: number | undefined;

  /**
   * 实际入库数总和
   */
  @Expose({
    name: \\"receive_sum_real\\"
  })
  @ValidateValue(/*NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型*/
  getDefaultValidNum())
  receiveSumReal: number | undefined;
}"
`;
